presto-main/src/main/java/com/facebook/presto/execution/QueryLimitMemoryRevokingScheduler.java

```diff
@@ -0,0 +1,173 @@
+
+    @Inject
+    public QueryLimitMemoryRevokingScheduler(
+            LocalMemoryManager localMemoryManager,
+            SqlTaskManager sqlTaskManager,
+            TaskManagementExecutor taskManagementExecutor)
+    {
+        this(
+                ImmutableList.copyOf(getMemoryPools(localMemoryManager)),
+                requireNonNull(sqlTaskManager, "sqlTaskManager cannot be null")::getQueryContext,
+                requireNonNull(taskManagementExecutor, "taskManagementExecutor cannot be null").getExecutor());
+        log.debug("Using QueryLimitMemoryRevokingScheduler spilling strategy");
+    }
+
+    @VisibleForTesting
+    QueryLimitMemoryRevokingScheduler(
+            List<MemoryPool> memoryPools,
+            Function<QueryId, QueryContext> queryContextSupplier,
+            ScheduledExecutorService taskManagementExecutor)
+    {
+        this.memoryPools = ImmutableList.copyOf(requireNonNull(memoryPools, "memoryPools is null"));
+        this.queryContextSupplier = requireNonNull(queryContextSupplier, "queryContextSupplier is null");
+        this.taskManagementExecutor = requireNonNull(taskManagementExecutor, "taskManagementExecutor is null");
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        registerPoolListeners();
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        memoryPools.forEach(memoryPool -> memoryPool.removeListener(memoryPoolListener));
+    }
+
+    @VisibleForTesting
+    void registerPoolListeners()
+    {
+        memoryPools.forEach(memoryPool -> memoryPool.addListener(memoryPoolListener));
+    }
+
+    private void onMemoryReserved(MemoryPool memoryPool, QueryId queryId, long queryTotalMemoryUse)
+    {
+        try {
+            QueryContext queryContext = queryContextSupplier.apply(queryId);
+            long maxTotalMemory = queryContext.getMaxTotalMemory();
+            if (queryTotalMemoryUse <= maxTotalMemory) {
+                return;
+            }
+
+            log.debug("Scheduling check for %s", queryId);
+            if (revocationRequestedByQuery.put(queryId, true) == null) {
+                scheduleRevoking(queryContext, maxTotalMemory);
+            }
+        }
+        catch (Throwable e) {
+            log.error(e, "Error when acting on memory pool reservation");
+        }
+    }
+
+    private void scheduleRevoking(QueryContext queryContext, long maxTotalMemory)
+    {
+        taskManagementExecutor.execute(() -> {
+            try {
+                revokeMemory(queryContext, maxTotalMemory);
+            }
+            catch (Throwable e) {
+                log.error(e, "Error requesting memory revoking");
+            }
+        });
+    }
+
+    private void revokeMemory(QueryContext queryContext, long maxTotalMemory)
+    {
+        QueryId queryId = queryContext.getQueryId();
+        MemoryPool memoryPool = queryContext.getMemoryPool();
+        while (!revocationRequestedByQuery.remove(queryId, false)) {
```

Comment:
I think this ordering works and removes duplicate revocation attempts.

What is the strategy around the node being blocked and needing revocation because it is out of memory? Say N queries each below the revocation limit, and N queries that have no revokable memory, but need to allocate? There will be no revocation until the N queries below the revocation limit hit the revocation limit for their particular query?

I know we talked about this, but I don't know what our final strategy was?

It seems to me like it's two parallel set of revocation conditions and actions. One is to revoke across queries and one is to revoke individual queries based on their own memory usage in isolation?

Reply:
yes, the TaskThresholdMemoryRevokingScheduler has the same problem.  The plan is to have a follow up PR that allows enabling both strategies.

Reply:
Ah OK. So we would have multiple strategies executed in sequence? I am not sure that is my favorite approach. Between DRY and having the implementation self-contained I feel like I would prefer the latter. Another reason is to reduce the repetition fetching values needed to determine what to do.

Another weirdness there is task execution order, but maybe they have a shared executor so it is a consistent order?

Reply:
I'll look more into what's the right way to combine them.